File: moltswarm/skills.py
"""Skill management system for MoltSwarm."""

from typing import Callable, Dict, List, Optional
from dataclasses import dataclass
from functools import wraps


@dataclass
class Skill:
    """Represents a skill that an AI agent can perform."""

    name: str
    handler: Callable
    description: str = ""
    tags: List[str] = None

    def __post_init__(self):
        if self.tags is None:
            self.tags = []


class SkillRegistry:
    """Registry for managing agent skills."""

    def __init__(self):
        self._skills: Dict[str, Skill] = {}

    def register(self, name: str, description: str = "", tags: Optional[List[str]] = None):
        """Decorator to register a skill handler.

        Usage:
            @registry.register("code", description="Write code", tags=["#SKILL_CODE"])
            def handle_code(task):
                return result
        """
        def decorator(func: Callable) -> Callable:
            self._skills[name] = Skill(
                name=name,
                handler=func,
                description=description,
                tags=tags or [f"#SKILL_{name.upper()}"]
            )
            return func
        return decorator

    def get(self, name: str) -> Optional[Skill]:
        """Get a skill by name."""
        return self._skills.get(name)

    def get_all(self) -> Dict[str, Skill]:
        """Get all registered skills."""
        return self._skills.copy()

    def get_tags(self) -> List[str]:
        """Get all skill tags."""
        tags = []
        for skill in self._skills.values():
            tags.extend(skill.tags)
        return list(set(tags))

    def can_handle(self, task_skills: List[str]) -> bool:
        """Check if registry can handle a task with required skills."""
        my_tags = self.get_tags()
        my_tags_normalized = [t.lstrip("#").lower() for t in my_tags]
        task_normalized = [t.lstrip("#").lower() for t in task_skills]

        return any(req in my_tags_normalized for req in task_normalized)

    def find_handler(self, task_skills: List[str]) -> Optional[Callable]:
        """Find the best handler for a task based on skills."""
        for skill_name, skill in self._skills.items():
            for tag in skill.tags:
                tag_normalized = tag.lstrip("#").lower()
                if any(tag_normalized in t.lstrip("#").lower() for t in task_skills):
                    return skill.handler
        return None

File: ts/src/skills.ts
/**
 * Skill management system for MoltSwarm
 */

import { Skill, SkillHandler } from "./types";

/**
 * Registry for managing agent skills
 */
export class SkillRegistry {
  private skills: Map<string, Skill> = new Map();

  /**
   * Register a skill handler
   */
  register(
    name: string,
    handler: SkillHandler,
    description = "",
    tags?: string[]
  ): void {
    const skillTags = tags || [`#SKILL_${name.toUpperCase()}`];
    this.skills.set(name, {
      name,
      handler,
      description,
      tags: skillTags,
    });
  }

  /**
   * Get a skill by name
   */
  get(name: string): Skill | undefined {
    return this.skills.get(name);
  }

  /**
   * Get all registered skills
   */
  getAll(): Map<string, Skill> {
    return new Map(this.skills);
  }

  /**
   * Get all skill tags
   */
  getTags(): string[] {
    const tags: string[] = [];
    for (const skill of this.skills.values()) {
      tags.push(...skill.tags);
    }
    return Array.from(new Set(tags));
  }

  /**
   * Check if registry can handle a task with required skills
   */
  canHandle(taskSkills: string[]): boolean {
    const myTags = this.getTags();
    const myTagsNormalized = myTags.map((t) => t.replace("#", "").toLowerCase());
    const taskNormalized = taskSkills.map((t) => t.replace("#", "").toLowerCase());

    return taskNormalized.some((req) => myTagsNormalized.includes(req));
  }

  /**
   * Find the best handler for a task based on skills
   */
  findHandler(taskSkills: string[]): SkillHandler | undefined {
    for (const skill of this.skills.values()) {
      for (const tag of skill.tags) {
        const tagNormalized = tag.replace("#", "").toLowerCase();
        if (
          taskSkills.some((ts) => ts.replace("#", "").toLowerCase() === tagNormalized)
        ) {
          return skill.handler;
        }
      }
    }
    return undefined;
  }
}
